src/libraries/System.IO.Pipelines/src/System/IO/Pipelines/StreamPipeReader.cs

```diff
@@ -308,37 +386,38 @@ private ReadOnlySequence<byte> GetCurrentReadOnlySequence()
             return new ReadOnlySequence<byte>(_readHead, _readIndex, _readTail, _readTail.End);
         }
 
-        private void AllocateReadTail()
+        private void AllocateReadTail(int? minimumSize = null)
         {
             if (_readHead == null)
             {
                 Debug.Assert(_readTail == null);
-                _readHead = AllocateSegment();
+                _readHead = AllocateSegment(minimumSize);
                 _readTail = _readHead;
             }
             else
             {
                 Debug.Assert(_readTail != null);
                 if (_readTail.WritableBytes < MinimumReadThreshold)
                 {
-                    BufferSegment nextSegment = AllocateSegment();
+                    BufferSegment nextSegment = AllocateSegment(minimumSize);
                     _readTail.SetNext(nextSegment);
                     _readTail = nextSegment;
                 }
             }
         }
 
-        private BufferSegment AllocateSegment()
+        private BufferSegment AllocateSegment(int? minimumSize = null)
         {
             BufferSegment nextSegment = CreateSegmentUnsynchronized();
 
+            var bufferSize = Math.Min(minimumSize ?? BufferSize, MaxBufferSize);
```

Comment:
Looking at this code now I think we have a bug (from one of my changes). We want to use 2 different maximums. If using the default shared pool, use MaxBufferSize, otherwise, we want to use Pool.MaxBufferSize. See https://github.com/dotnet/runtime/blob/cbe195eef04135e29bf0ba164b86f4b9c10b0231/src/libraries/System.IO.Pipelines/src/System/IO/Pipelines/Pipe.cs#L214-L233

Reply:
Fixed.

Reply:
I think this logic needs more work:
- If there is a pool configured, use it unless the requested buffer size is too big
- Otherwise use the array pool

Reply:
I think I have understood your remark this time... is it better now?

Reply:
Yep